sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/TooManyParameters.cs

```diff
@@ -0,0 +1,118 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.VisualBasic;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class TooManyParameters : TooManyParametersBase<SyntaxKind, ParameterListSyntax>
+    {
+        protected override GeneratedCodeRecognizer GeneratedCodeRecognizer => VisualBasicGeneratedCodeRecognizer.Instance;
+        protected override SyntaxKind[] SyntaxKinds => new SyntaxKind[] { SyntaxKind.ParameterList };
+        protected override DiagnosticDescriptor Rule => rule;
+        protected override Dictionary<SyntaxKind, string> Mapping => mapping;
+        protected override SyntaxKind ParentType(ParameterListSyntax parameterList) => parameterList.Parent.Kind();
+        protected override int CountParameters(ParameterListSyntax parameterList) => parameterList.Parameters.Count;
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+
+        protected override bool CanBeChanged(SyntaxNode node, SemanticModel semanticModel)
+        {
+            var declaredSymbol = semanticModel.GetDeclaredSymbol(node);
+            var symbol = semanticModel.GetSymbolInfo(node).Symbol;
+
+            if (node.IsAnyKind(LambdaHeaders))
+            {
+                return true;
+            }
+
+            if (declaredSymbol == null && symbol == null)
+            {
+                // no information
+                return false;
+            }
+
+            if (symbol != null)
+            {
+                // Not a declaration, such as Action
+                return true;
+            }
+
+            if ((node as SubNewStatementSyntax)?.ParameterList?.Parameters.Count is int parameterCount &&
+                parameterCount > Maximum &&
+                node.Parent is ConstructorBlockSyntax constructorBlock &&
+                ContainsMyBaseNewInvocation(constructorBlock, Maximum))
+            {
+                return false;
+            }
+
+            if (declaredSymbol.IsExtern &&
+                declaredSymbol.IsStatic &&
+                declaredSymbol.GetAttributes(KnownType.System_Runtime_InteropServices_DllImportAttribute).Any())
+            {
+                // P/Invoke method is defined externally.
+                // Do not raise
+                return false;
+            }
+
+            return declaredSymbol.GetOverriddenMember() == null &&
+                   declaredSymbol.GetInterfaceMember() == null;
+        }
+
+        private static readonly DiagnosticDescriptor rule =
```

Comment:
Move this one upper in the class so that C# and VB.NET classes are alike in term of ordering.

Reply:
This comment wasn't fixed. We usually put the fields on top of the class (at least before method definitions).

Reply:
Ok. I think all these conventions have to be formalized. Smth like:
- Properties sorted by visibility
- Fields sorted by visibility
- Methods sorted by visibility